src/main/java/com/marginallyclever/makelangelo/makeArt/io/vector/LoadScratch2.java

```diff
@@ -154,24 +105,81 @@ public Turtle load(InputStream in) throws Exception {
     			String firstType = (String)scriptHead;
 				if(firstType.equals("whenGreenFlag")) {
 					logger.debug("Found green flag script");
-					if( greenFlagScript != null ) {
-						throw new Exception("More than one green flag script");
-					}
-					greenFlagScript = scripts02;
-				} else if(firstType.equals("whenIReceive")) {
+					parseScratchCode(scripts02);
+				}
+			}
+		}
+	}
+
+	private void readAllSubroutines(JSONArray scripts) {
+		// extract known elements and convert them to turtle commands.
+		Iterator<?> scriptIter = scripts.iterator();
+		
+		while( scriptIter.hasNext() ) {
+			JSONArray scripts0 = (JSONArray)scriptIter.next();
+			if( scripts0==null ) continue;
+			JSONArray scripts02 = (JSONArray)scripts0.get(2);
+			if( scripts02==null || scripts02.length()==0 ) continue;
+			// look inside the script at the first block, which gives the type.
+			Object scriptContents = scripts02.get(0);
+			if( ( scriptContents instanceof JSONArray ) ) {
+				String firstType = ((JSONArray)scriptContents).getString(0);
+				if(firstType.equals("whenIReceive")) {
 	    			String firstName = (String)((JSONArray)scriptContents).get(1);
 					logger.debug("Found subroutine {}", firstName);
 	    			subRoutines.put(firstName, ((JSONArray)scriptContents));
 				}
 			}
 		}
+	}
 
-		// actual code begins here.
-		parseScratchCode(greenFlagScript);
+	private JSONArray getScripts(JSONObject tree) throws Exception {
+		Iterator<?> childIter = tree.getJSONArray("children").iterator();
+		while( childIter.hasNext() ) {
+			JSONObject child = (JSONObject)childIter.next();
+			if(child.has("scripts")) {
+				return (JSONArray)child.get("scripts");
+			}
+		}
+		throw new Exception("JSON node 'scripts' missing.");
+	}
+
+	private JSONObject getTreeFromInputStream(InputStream in) throws FileNotFoundException, IOException {
+		File tempZipFile = extractProjectJSON(in);
 		
-		logger.debug("finished scripts");
+        logger.debug("Parsing JSON file...");
+        JSONTokener tokener = new JSONTokener(tempZipFile.toURI().toURL().openStream());
+        JSONObject tree = new JSONObject(tokener);
 
-		return turtle;
+		tempZipFile.delete();
+		
+		return tree;
+	}
+
+	private File extractProjectJSON(InputStream in) throws FileNotFoundException, IOException {
+    	logger.debug("Searching for project.json...");
+		ZipInputStream zipInputStream = new ZipInputStream(in);
+		ZipEntry entry;
+		while((entry = zipInputStream.getNextEntry())!=null) {
+	        if(entry.getName().equals(PROJECT_JSON) ) {
+	        	logger.debug("Found.");
+	        	
+		        // read buffered stream into temp file.
+	        	File tempZipFile = File.createTempFile("project", "json");
+	        	tempZipFile.setReadable(true);
+	        	tempZipFile.setWritable(true);
+	        	tempZipFile.deleteOnExit();
+		        FileOutputStream fos = new FileOutputStream(tempZipFile);
+	    		byte[] buffer = new byte[2048];
```

Comment:
The code indentation is not homogeneous: there are 4 extra spaces in thoses lines

Reply:
fixed and also in LoadScratch2